src/frontend/src/components/tables/InvenTreeTable.tsx

```diff
@@ -0,0 +1,347 @@
+        // Keep a shadow copy of the state variable, so that we can update it asynchronously
+        latestSearchTerm = searchTerm;
+
+        loadHiddenColumns();
+        loadActiveFilters();
+    }, [searchTerm]);
+
+    // Load list of hidden columns from local storage
+    function loadHiddenColumns() {
+        hiddenColumns = JSON.parse(localStorage.getItem(`hidden-table-columns-${tableKey}`) || '[]');
+
+        // Update column visibility
+        let newColumns = [...dataColumns];
+
+        for (let idx = 0; idx < newColumns.length; idx++) {
+            newColumns[idx].hidden = hiddenColumns.includes(newColumns[idx].accessor);
+        }
+
+        setDataColumns(newColumns);
+    }
+
+    // Load list of active filters from local storage
+    function loadActiveFilters() {
+        // TODO
+    }
+
+    // Callback when search term is updated
+    function updateSearchTerm(term: string) {
+        term = term.trim();
+        // Ignore identical search terms
+        if (term == searchTerm) return;
+
+        setSearchTerm(term);
+        latestSearchTerm = term;
+        refetch();
+    }
+
+    // Data download callback
+    function downloadData(fileFormat: string) {
+        console.log("download data: " + fileFormat);
+    }
+
+    // Data Sorting
+    const [sortStatus, setSortStatus] = useState<DataTableSortStatus>({ columnAccessor: defaultSortColumn, direction: 'asc' });
+
+    // Return the ordering parameter
+    function getOrderingTerm() {
+        let key = sortStatus.columnAccessor;
+
+        // Sorting column not specified
+        if (key == '') {
+            return '';
+        }
+
+        // Find matching column:
+        // If column provides custom ordering term, use that
+        let column = dataColumns.find((col) => col.accessor == key);
+
+        if (column && column.ordering) {
+            return column.ordering;
+        } else {
+            return key;
+        }
+    }
+
+    // Missing records text (based on server response)
+    const [missingRecordsText, setMissingRecordsText] = useState<string>(noRecordsText);
+
+    // Data selection
+    const [selectedRecords, setSelectedRecords] = useState<any[]>([]);
+
+    function onSelectedRecordsChange(records: any[]) {
+        setSelectedRecords(records);
+    }
+
+    const handleSortStatusChange = (status: DataTableSortStatus) => {
+        setPage(1);
+        setSortStatus(status);
+    };
+
+    // Callback when column visibility is toggled
+    function toggleColumn(columnName: string) {
+        let newColumns = [...dataColumns];
+
+        let hiddenColumns = [];
+
+        // Toggle selected column
+        for (let idx = 0; idx < newColumns.length; idx++) {
+
+            if (newColumns[idx].accessor == columnName) {
+                newColumns[idx].hidden = !newColumns[idx].hidden;
+            }
+
+            if (newColumns[idx].hidden) {
+                hiddenColumns.push(newColumns[idx].accessor);
+            }
+        }
```

Comment:
You could utilize the es6 array map, filter, reduce functions here. They become pretty handy and look much more modern than such cpp style for loops.